### PR TITLE
Force storage of incomplete analyses

### PR DESCRIPTION
**Pull request by ivadym**
### Added
- Force flag to store command,
- `cli_options.py` constants file for shared click options/arguments.

Hermes PR:
- https://github.com/Clinical-Genomics/hermes/pull/113

Next steps (https://github.com/Clinical-Genomics/cg/issues/2702):
- Allow delivery/upload of analyses missing metrics deliverables file.
- Add a comment column to analysis table and include it in the delivery report for failed analyses

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- `cg workflow <workflow> store <case_id>`

### Expected test outcome

- [x] Check that the case storage fails for incomplete analysis
    - [x] Balsamic
    - [x] MIP
    - [x] Nextflow pipelines (Rnafusion)
- [x] Check that the case storage succeeds for incomplete analysis with the `--force` flag
    - [x] Balsamic
    - [x] MIP
    - [x] Nextflow pipelines (Rnafusion)

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Review by ChrOertlin — APPROVED** [review]
Looks good, mostly formatting changes and addition of force flags.

This will only be a manual option, right?

**Comment by ivadym** [comment]
> Looks good, mostly formatting changes and addition of force flags.
> 
> This will only be a manual option, right?

Yes, exactly. The analysis will be stored manually and will remain in the failed step, so it won't be picked up by automation

File: cg/cli/generate/report/base.py [hunk 1]
```diff
@@ -51,7 +50,7 @@ def generate_delivery_report(
     # Dry run: prints the HTML report to console
     if dry_run:
         delivery_report_html: str = report_api.create_delivery_report(
-            case_id, analysis_date, force_report
+            case_id=case_id, analysis_date=analysis_date, force=force
```
Comment (henrikstranneheim) [inline cg/cli/generate/report/base.py:53]: 👍

File: cg/cli/workflow/balsamic/base.py [hunk 2]
```diff
@@ -128,16 +128,19 @@ def run(
 @balsamic.command("report-deliver")
 @ARGUMENT_CASE_ID
 @DRY_RUN
+@FORCE
 @click.pass_obj
-def report_deliver(context: CGConfig, case_id: str, dry_run: bool):
+def report_deliver(context: CGConfig, case_id: str, dry_run: bool, force: bool):
     """Create a housekeeper deliverables file for given CASE ID"""
```
Comment (henrikstranneheim) [inline cg/cli/workflow/balsamic/base.py:134]:
```suggestion
    """Create a Housekeeper deliverables file for given case ID"""
```

File: cg/cli/workflow/nf_analysis.py [hunk 3]
```diff
@@ -239,12 +240,13 @@ def metrics_deliver(context: CGConfig, case_id: str, dry_run: bool) -> None:
 @click.command("report-deliver")
 @ARGUMENT_CASE_ID
 @DRY_RUN
+@FORCE
 @click.pass_obj
-def report_deliver(context: CGConfig, case_id: str, dry_run: bool) -> None:
-    """Create a Housekeeper deliverables file for given case id."""
+def report_deliver(context: CGConfig, case_id: str, dry_run: bool, force: bool) -> None:
+    """Create a Housekeeper deliverables file for a given case ID."""
```
Comment (henrikstranneheim) [inline cg/cli/workflow/nf_analysis.py:246]: Add Raises

File: cg/cli/workflow/nf_analysis.py [hunk 4]
```diff
@@ -253,12 +255,13 @@ def report_deliver(context: CGConfig, case_id: str, dry_run: bool) -> None:
 @click.command("store-housekeeper")
 @ARGUMENT_CASE_ID
 @DRY_RUN
+@FORCE
 @click.pass_obj
-def store_housekeeper(context: CGConfig, case_id: str, dry_run: bool) -> None:
+def store_housekeeper(context: CGConfig, case_id: str, dry_run: bool, force: bool) -> None:
     """Store a finished nf-analysis in Housekeeper and StatusDB."""
```
Comment (henrikstranneheim) [inline cg/cli/workflow/nf_analysis.py:261]:
```suggestion
    """Store a finished NF-analysis in Housekeeper and StatusDB."""
```
Add raises

File: cg/cli/workflow/nf_analysis.py [hunk 5]
```diff
@@ -267,13 +270,14 @@ def store_housekeeper(context: CGConfig, case_id: str, dry_run: bool) -> None:
 @click.command("store")
 @ARGUMENT_CASE_ID
 @DRY_RUN
+@FORCE
 @click.pass_context
-def store(context: click.Context, case_id: str, dry_run: bool) -> None:
+def store(context: click.Context, case_id: str, dry_run: bool, force: bool) -> None:
     """Generate deliverable files for a case and store in Housekeeper if they
     pass QC metrics checks."""
```
Comment (henrikstranneheim) [inline cg/cli/workflow/nf_analysis.py:277]: Add raises

File: cg/meta/report/report_api.py [hunk 6]
```diff
@@ -40,26 +55,24 @@ def __init__(self, config: CGConfig, analysis_api: AnalysisAPI):
         super().__init__(config=config)
         self.analysis_api: AnalysisAPI = analysis_api
 
-    def create_delivery_report(
-        self, case_id: str, analysis_date: datetime, force_report: bool
-    ) -> str:
+    def create_delivery_report(self, case_id: str, analysis_date: datetime, force: bool) -> str:
         """Generates the html contents of a delivery report."""
```
Comment (henrikstranneheim) [inline cg/meta/report/report_api.py:59]:
```suggestion
        """Generates the HTML contents of a delivery report."""
```

File: cg/meta/workflow/analysis.py [hunk 7]
```diff
@@ -212,10 +212,12 @@ def has_case_only_exome_samples(self, case_id: str) -> bool:
         application_type: str = self.get_case_application_type(case_id)
         return application_type == AnalysisType.WHOLE_EXOME_SEQUENCING
 
-    def upload_bundle_housekeeper(self, case_id: str, dry_run: bool = False) -> None:
+    def upload_bundle_housekeeper(
+        self, case_id: str, dry_run: bool = False, force: bool = False
+    ) -> None:
         """Storing bundle data in Housekeeper for CASE_ID"""
```
Comment (henrikstranneheim) [inline cg/meta/workflow/analysis.py:218]:
```suggestion
        """Storing bundle data in Housekeeper for case ID"""
```

File: cg/meta/workflow/analysis.py [hunk 8]
```diff
@@ -294,13 +296,14 @@ def _get_order_id_from_case_id(self, case_id) -> int:
         case: Case = self.status_db.get_case_by_internal_id(case_id)
         return case.latest_order.id
 
-    def get_hermes_transformed_deliverables(self, case_id: str) -> dict:
+    def get_hermes_transformed_deliverables(self, case_id: str, force: bool = False) -> dict:
         return self.hermes_api.create_housekeeper_bundle(
```
Comment (henrikstranneheim) [inline cg/meta/workflow/analysis.py:300]: Would be nice with a doc str here.

File: cg/meta/workflow/nf_analysis.py [hunk 9]
```diff
@@ -759,31 +759,31 @@ def metrics_deliver(self, case_id: str, dry_run: bool):
...
             f"Writing deliverables file in {self.get_deliverables_file_path(case_id=case_id).as_posix()}"
         )
 
-    def store_analysis_housekeeper(self, case_id: str, dry_run: bool = False) -> None:
-        """Store a finished nextflow analysis in Housekeeper and StatusDB"""
-
+    def store_analysis_housekeeper(
+        self, case_id: str, dry_run: bool = False, force: bool = False
+    ) -> None:
+        """Store a finished nextflow analysis in Housekeeper and StatusDB."""
```
Comment (henrikstranneheim) [inline cg/meta/workflow/nf_analysis.py:781]: Add raises

**Review by henrikstranneheim — APPROVED** [review]
Great! 👍

**Comment by ivadym** [comment]
Screenshot 2024-06-17 at 08 19 26

**Pull request merged**